### PR TITLE
[Spark-Doris-Connector][Bug-Fix] Resolve deserialize exception when Spark Doris Connector in aync deserialize mode

### LINKED ISSUE #5335: [Bug] ScalaValueReader throw TApplicationException: get_next failed: out of sequence response: expected 21 but got 20
**Describe the bug**

When I use spark on doris with config `doris.deserialize.arrow.async=true`，ScalaValueReader throw TApplicationException: get_next failed: out of sequence response: expected 21 but got 20.

And then, ScalaValueReader will **block forever**

This is because, when `doris.deserialize.arrow.async=true`, `get_next()` will invoke in **asyncThread**，but `close()` will invoke in **current thread**, the sequence will encounter race condition 

**To Reproduce**
run
``` scala
val sparkSettings = new SparkSettings(sparkConf)
sparkSettings.setProperty("doris.deserialize.arrow.async", "true")
val partitions = RestService.findPartitions(sparkSettings, logger)
val reader = new ScalaDorisRowValueReader(partitions.get(0), partitions.get(0).settings())
while (reader.hasNext) {
  val nextBatch = reader.next // invoke in asyncThread
}
reader.close() // invoke in current thread
```

**Expected behavior**
No exception

**Screenshots**
image

**Desktop (please complete the following information):**
None

**Smartphone (please complete the following information):**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by 924060929**
## Proposed changes

fix #5335

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [] I have added tests that prove my fix is effective or that my feature works
- [] If these changes need document changes, I have updated the document
- [] Any dependent changes have been merged

## Further comments
None

### CODE REVIEW

**Comment by 924060929** [comment]
Gentle ping @wuyunfeng

**Comment by wuyunfeng** [comment]
@924060929 This is a nice capture. I am curious about whether  the lock is necessary when running in sync mode?

**Comment by 924060929** [comment]
> @924060929 This is a nice capture. I am curious about whether the lock is necessary when running in sync mode （default mode）?

Unnecessary, just for simplify. Synchronized has little effect on performance in sync mode.

Shall we remove synchronized in sync mode?

**Comment by wuyunfeng** [comment]
@924060929 I suggest that there can be such an optimization

**Comment by 924060929** [comment]
@wuyunfeng OK, I will add this optimization soon

**Comment by 924060929** [comment]
> @924060929 I suggest that there can be such an optimization

@wuyunfeng done

**Comment by wuyunfeng** [comment]
I just using `Java` describe the implementation which I would prefer:

First define a `NoOpLock`:

```
protected static final class NoOpLock implements Lock {

        @Override
        public void lock() {
        }

        @Override
        public void lockInterruptibly() throws InterruptedException {
        }

        @Override
        public boolean tryLock() {
            return true;
        }

        @Override
        public boolean tryLock(long time, TimeUnit unit) throws InterruptedException {
            return true;
        }

        @Override
        public void unlock() {
        }

        @Override
        public Condition newCondition() {
            throw new UnsupportedOperationException("NoOpLock can't provide a condition");
        }
    }
``` 
We just define a `private final Lock lock = new NoOpLock();`

If using `async` mode,  let the lock using the real `Lock`, If not, we can replace the `lock` with `NoOpLock`.

If we do this, the first code is more elegant, and second, we can avoid excessive branch switching.

 I think your approach is fine, would you like to think about my suggestion

**Comment by 924060929** [comment]
@wuyunfeng OK, first I think the same solution as you note, but don't know where the place the NoOpLock, we can add a inner class, or add some public class in any package, or any third-party package has same implement?

So I add the if condition without Lock implement.

Can you give more advice?

**Comment by wuyunfeng** [comment]
I think you can just use a inner class

**Comment by 924060929** [comment]
> I think you can just use a inner class

done

**Review by wuyunfeng — DISMISSED** [review]
To be honest, I don’t like this function name  `lockClient` too much, but this  looks good to me

File: extension/spark-doris-connector/src/main/scala/org/apache/doris/spark/rdd/ScalaValueReader.scala [hunk 1]
```diff
@@ -46,6 +47,9 @@ class ScalaValueReader(partition: PartitionDefinition, settings: Settings) {
   protected val logger = Logger.getLogger(classOf[ScalaValueReader])
 
   protected val client = new BackendClient(new Routing(partition.getBeAddress), settings)
+  protected val clientLock =
+    if (deserializeArrowToRowBatchAsync) new ReentrantLock()
```
Comment (Youngwb) [inline extension/spark-doris-connector/src/main/scala/org/apache/doris/spark/rdd/ScalaValueReader.scala:51]: I have a question， deserializeArrowToRowBatchAsync is init below,  whether the use is valid here

Comment (924060929) [reply extension/spark-doris-connector/src/main/scala/org/apache/doris/spark/rdd/ScalaValueReader.scala:51]: Is valid, equals to c++:
```c++
bool deserializeArrowToRowBatchAsync() {
  return ...
}
```

Comment (924060929) [reply extension/spark-doris-connector/src/main/scala/org/apache/doris/spark/rdd/ScalaValueReader.scala:51]: My fault, `protected var deserializeArrowToRowBatchAsync` should be `protected def deserializeArrowToRowBatchAsync`

**Comment by 924060929** [comment]
@Youngwb already fix bug which use deserializeArrowToRowBatchAsync before initialize

**Review by Youngwb — APPROVED** [review]
+1

**Review by wuyunfeng — APPROVED** [review]
LGTM

**Comment by wuyunfeng** [comment]
@924060929 I have merge this into master. Thanks

**Pull request merged**